source/common/tls/tls_certificate_selector_impl.cc

```diff
@@ -0,0 +1,57 @@
+#include "source/common/tls/tls_certificate_selector_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+Ssl::SelectionResult
+TlsCertificateSelectorImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello,
+                                             Ssl::CertSelectionCallbackPtr) {
+  auto selection_ctx = cb_.lock();
+  if (selection_ctx == nullptr) {
+    IS_ENVOY_BUG("ContextSelectionCallback is gone early");
+    return {Ssl::SelectionResult::SelectionStatus::AbortHandshake, nullptr, false};
+  }
+
+  auto server_ctx = std::dynamic_pointer_cast<Extensions::TransportSockets::Tls::ServerContextImpl>(
+      selection_ctx);
+
+  absl::string_view sni = absl::NullSafeStringView(
+      SSL_get_servername(ssl_client_hello->ssl, TLSEXT_NAMETYPE_host_name));
+  const bool client_ecdsa_capable = server_ctx->isClientEcdsaCapable(ssl_client_hello);
+  const bool client_ocsp_capable = server_ctx->isClientOcspCapable(ssl_client_hello);
+
+  auto [selected_ctx, ocsp_staple_action] =
+      server_ctx->findTlsContext(sni, client_ecdsa_capable, client_ocsp_capable, nullptr);
```

Comment:
Should the body of that function move to this class? And it's data (name map of certs)?